File: bots/intgrah/v54.7.4/builder/tasks/offense/turret_around_harvester.py
"""Place gunner / sentinel turrets adjacent to a vulnerable enemy
harvester, capping at 2 gunners + 1 sentinel per harvester. Builder
must be on empty terrain at d²==1 of the harvester. Sequence: random
step-off (so the build tile is unblocked), pick chain-facing for the
gunner via `gunner_chain_facing` (ray must end on enemy transport),
sentinel facing toward enemy core, pave a road on the build tile, then
scout toward enemy. Atomic — one task call does all four steps.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, override

from building import BuildingGunner, BuildingSentinel
from cambc import EntityType, Environment
from util.directions import DIR4

from builder.helpers import move_random, try_place
from builder.tasks.offense.helpers import (
    gunner_chain_facing,
    is_allied_transport,
    pick_harvester_target,
    scout_toward_enemy,
    vulnerable_harvesters,
)
from builder.tasks.rejected import Reason, TaskRejectedError

if TYPE_CHECKING:
    from cambc import Controller

    from builder import Builder


class TaskRejectedNotAdjacentEmptyError(TaskRejectedError):
    @override
    def reason(self) -> Reason:
        return "not on empty terrain cardinal to a vulnerable harvester"


def turret_around_harvester(self: Builder, ct: Controller) -> None:
    # Preconditions: standing on empty (non-enemy-building) terrain at
    # d²==1 of a vulnerable harvester, not on allied transport.
    vulnerable = vulnerable_harvesters(self)
    if not vulnerable:
        raise TaskRejectedNotAdjacentEmptyError
    target = pick_harvester_target(self, vulnerable)
    if self.my_pos.distance_squared(target) != 1:
        raise TaskRejectedNotAdjacentEmptyError
    if is_allied_transport(self, self.my_pos):
        raise TaskRejectedNotAdjacentEmptyError
    if self.is_enemy_building(self.my_pos):
        raise TaskRejectedNotAdjacentEmptyError

    # Preserve original's exact sequence:
    #   1. move_random() — required so we can place turrets on the tile
    #      we just vacated (ct.can_build rejects a tile with a builder
    #      standing on it).
    #   2. compute facing direction toward enemy_core, rotate right if
    #      that collides with direction-to-target.
    #   3. count existing gunners/sentinels cardinal to target.
    #   4. place gunner at build_position if caps allow and chain
    #      facing exists.
    #   5. place sentinel at build_position if caps allow and target is
    #      on Ti ore.
    #   6. pave build_position with a road if possible.
    #   7. scout_toward_enemy.
    build_position = self.my_pos
    enemy_core = self.en_core_guess
    move_random(self, ct)
    direction = build_position.direction_to(enemy_core)
    if direction == build_position.direction_to(target):
        direction = direction.rotate_right()

    n_gunner = 0
    n_sentinel = 0
    for d in DIR4:
        n = target.add(d)
        if not self.in_bounds(n):
            continue
        nb = self.get_building(n)
        if nb is None or nb.team != self.my_team:
            continue
        if isinstance(nb, BuildingGunner):
            n_gunner += 1
        elif isinstance(nb, BuildingSentinel):
            n_sentinel += 1

    if n_gunner < 2:
        gdir = gunner_chain_facing(self, build_position)
        if gdir is not None:
            try_place(self, ct, EntityType.GUNNER, build_position, gdir)

    if n_sentinel == 0 and self.get_env(target) == Environment.ORE_TITANIUM:
        try_place(self, ct, EntityType.SENTINEL, build_position, direction)

    if ct.can_build_road(build_position):
        ct.build_road(build_position)
    scout_toward_enemy(self, ct)

File: bots/rs/v56/src/builder/tasks/offense/turret_around_harvester.rs
//! Place gunner / sentinel turrets adjacent to a vulnerable enemy
//! harvester, capping at 2 gunners + 1 sentinel per harvester.

use cambc::{BuildExtra, Controller, ControllerApi, Direction, EntityType, Environment, Position};

use crate::builder::Builder;
use crate::builder::helpers::{can_afford, move_random, try_place};
use crate::builder::tasks::offense::helpers::{
    gunner_chain_facing, is_allied_transport, pick_harvester_target, scout_toward_enemy,
    vulnerable_harvesters,
};
use crate::builder::tasks::econ::infrastructure::place_gunner::{is_resource_building, safe_facing};
use crate::builder::tasks::rejected::{TaskRejected, TaskResult};
use crate::util::directions::{DIR4, is_cardinal};

/// Snap the unit vector from `src` to `dst` to the nearest 45-degree direction.
const fn direction_to(src: Position, dst: Position) -> Direction {
    let dx = dst.x - src.x;
    let dy = dst.y - src.y;
    if dx == 0 && dy == 0 {
        return Direction::Centre;
    }
    let adx = pyrust::abs!(dx);
    let ady = pyrust::abs!(dy);
    if adx * 5 < ady * 2 {
        return if dy < 0 {
            Direction::North
        } else {
            Direction::South
        };
    }
    if ady * 5 < adx * 2 {
        return if dx < 0 {
            Direction::West
        } else {
            Direction::East
        };
    }
    match (pyrust::signum!(dx), pyrust::signum!(dy)) {
        (1, -1) => Direction::Northeast,
        (1, 1) => Direction::Southeast,
        (-1, 1) => Direction::Southwest,
        (-1, -1) => Direction::Northwest,
        _ => Direction::Centre,
    }
}

const fn rotate_right(d: Direction) -> Direction {
    match d {
        Direction::North => Direction::Northeast,
        Direction::Northeast => Direction::East,
        Direction::East => Direction::Southeast,
        Direction::Southeast => Direction::South,
        Direction::South => Direction::Southwest,
        Direction::Southwest => Direction::West,
        Direction::West => Direction::Northwest,
        Direction::Northwest => Direction::North,
        Direction::Centre => Direction::Centre,
    }
}

pub fn turret_around_harvester(self_: &mut Builder, ct: &mut Controller<'_>) -> TaskResult {
    let vulnerable = vulnerable_harvesters(self_);
    if pyrust::vec::is_empty!(vulnerable) {
        return Some(TaskRejected::new("no vulnerable harvester"));
    }
    let target = pick_harvester_target(self_, &vulnerable);
    if self_.my_pos.distance_squared(target) != 1 {
        return Some(TaskRejected::new(
            "not cardinally adjacent to a vulnerable harvester",
        ));
    }
    if is_allied_transport(self_, self_.my_pos) {
        return Some(TaskRejected::new(
            "on allied transport cardinal to a vulnerable harvester",
        ));
    }
    if self_.is_enemy_building(self_.my_pos) {
        return Some(TaskRejected::new(
            "on enemy building cardinal to a vulnerable harvester",
        ));
    }

    let build_position = self_.my_pos;
    let enemy_core = self_.en_core_guess;
    let mut direction = direction_to(build_position, enemy_core);
    if direction == direction_to(build_position, target) {
        direction = rotate_right(direction);
    }

    let mut n_gunner = 0;
    let mut n_sentinel = 0;
    for d in DIR4 {
        let n = target.add(d);
        if !self_.in_bounds(n) {
            continue;
        }
        let Some((nk, nt)) = self_.get_building(n) else {
            continue;
        };
        if nt != self_.my_team {
            continue;
        }
        if nk == EntityType::Gunner {
            n_gunner += 1;
        } else if nk == EntityType::Sentinel {
            n_sentinel += 1;
        }
    }

    let gdir = if n_gunner < 2 {
        gunner_chain_facing(self_, build_position)
    } else {
        None
    };
    let want_gunner = pyrust::is_some!(gdir) && can_afford(self_, EntityType::Gunner);
    let sentinel_dir_ok = !is_cardinal(direction)
        || !self_.in_bounds(build_position.add(direction))
        || !is_resource_building(self_.kind_at(build_position.add(direction)))
        || self_.team_at(build_position.add(direction)) != Some(self_.my_team);
    let want_sentinel = sentinel_dir_ok
        && n_sentinel == 0
        && self_.get_env(target) == Some(Environment::OreTitanium)
        && can_afford(self_, EntityType::Sentinel);

    let mut placed = false;
    if want_gunner || want_sentinel {
        move_random(self_, ct);
        if want_gunner {
            let gd = safe_facing(self_, build_position, pyrust::unwrap!(gdir));
            placed = try_place(
                self_,
                ct,
                EntityType::Gunner,
                build_position,
                BuildExtra::Direction(gd),
                true,
            );
        }
        if !placed && want_sentinel {
            placed = try_place(
                self_,
                ct,
                EntityType::Sentinel,
                build_position,
                BuildExtra::Direction(direction),
                true,
            );
        }
    }

    if !placed {
        if pyrust::unwrap!(ct.can_build_road(build_position)) {
            pyrust::unwrap!(ct.build_road(build_position));
        } else {
            return Some(TaskRejected::new("couldn't place a sentinel or road"));
        }
    }

    scout_toward_enemy(self_, ct);
    None
}
